src/proxy/moralis.rs

```diff
@@ -0,0 +1,330 @@
+use crate::address_status::{AddressStatus, AddressStatusOperations};
+use crate::ctx::{AppConfig, ProxyRoute};
+use crate::db::Db;
+use crate::http::{
+    insert_jwt_to_http_header, response_by_status, APPLICATION_JSON, X_FORWARDED_FOR,
+};
+use crate::proxy::{remove_hop_by_hop_headers, X_AUTH_PAYLOAD};
+use crate::rate_limiter::RateLimitOperations;
+use crate::sign::{SignOps, SignedMessage};
+use crate::{log_format, GenericResult};
+use hyper::header::{HeaderName, HeaderValue};
+use hyper::http::uri::PathAndQuery;
+use hyper::{header, Body, Request, Response, StatusCode, Uri};
+use hyper_tls::HttpsConnector;
+use serde::{Deserialize, Serialize};
+use std::net::SocketAddr;
+use std::str::FromStr;
+use url::Url;
+
+/// Represents a payload for HTTP GET requests, specifically parsed for the Moralis API within the proxy.
+/// This struct contains the destination URL that the proxy will forward the GET request to, ensuring correct service routing.
```

Comment:
Can you explain the motivation behind this? Why are we dynamically taking the target proxy address?

Reply:
What I see is that only the endpoint changes not the whole URL. The proxy shouldn't need to look for this data in the request. Clients should only need to provide signed message in the header and the rest must be as same as like there is not proxy/KDP on the receiver side.

For example, with the configuration:

```json
"inbound_route": "ETH",
"outbound_route": "quicknode.com/user45135134"
```

KDP will handle requests like this:

- `komodo-defi-proxy.com/ETH/some/specific/endpoint` is forwarded to `quicknode.com/user45135134/some/specific/endpoint`
- `komodo-defi-proxy.com/ETH/1` is forwarded to `quicknode.com/user45135134/1`
- `komodo-defi-proxy.com/ETH` is forwarded to `quicknode.com/user45135134`

There's no additional processing required for the komodo-defi-framework or clients. They just need to add a signed message to the header. Everything else works the same as using the services directly without a proxy (which is typical for proxies by their nature).

Reply:
> There's no additional processing required for the komodo-defi-framework or clients. They just need to add a signed message to the header. Everything else works the same as using the services directly without a proxy (which is typical for proxies by their nature).

Not the same in our case, even in your example you are adding additional path segment "ETH". 
Due to `inbound/outbound` routes GET requests dont work directly. It will work for POST yes.

I tried to keep functionality as generic as possible. To find `outbound_rout`, you just need to provide correct path to proxy uri, eg `/ethereum`, `/nft` etc. It depends what `inbound routes` are set in app config
![Screenshot 2024-05-21 at 12 44 38](https://github.com/KomodoPlatform/komodo-defi-proxy/assets/52405288/306b687f-7135-4e20-9241-236568321da7)

Actually your examples above confuses me. Did you take into account how `get_proxy_route_by_inbound` will work to find `outbound_route` by `inbound_route` for POST and GET methods and for different endpoints?

Just note, we use three moralis endpoints:
- Get NFT transfers by wallet
GET https://deep-index.moralis.io/api/v2.2/:address/nft/transfers
- Get NFTs by wallet
GET https://deep-index.moralis.io/api/v2.2/:address/nft
- Get NFT metadata
GET https://deep-index.moralis.io/api/v2.2/nft/:address/:token_id

I suggest to send GET moralis requests to proxy to `https://komodoproxy.com/nft` Url and I suggest to provide auth message in `X-Auth-Payload` header + target url in `X-Target-Url` header.  
If we already ask for signed message in header, why not ask target url. So proxy layer will easily find `outbound_route` by `inbound_route` and also without problems will parse signed message and target url (Note that in target url we still need to change `https://komodoproxy.com` to base url in app configuration).

If we go with your approach:
Lets just see the real example.

1. **KDF Sends Request**:
   - KDF sends a GET request to `https://komodoproxy.com/nft-proxy/api/v2.2/0x1f9090aaE28b8a3dCeaDf281B0F12828e676c326/nft/transfers?chain=eth&format=decimal&order=DESC` with the `X-Auth-Payload` header.

2. **KDP Processes Request**:
   - KDP checks if the request method is GET.
   - If it is, KDP verifies if the first segment in the path exists.
   - If the first segment exists and equals `nft-proxy`, KDP finds the outbound route based on this segment.
   - For a more general implementation, KDP can iterate through all proxy routes using the first path segment as the inbound route without specifically checking for `nft-proxy`.

3. **Modify Request URL**:
   - Once the outbound route is found, KDP parses the header and modifies the request URL:
     - Removes the `nft-proxy` segment from the path.
     - Changes the base URL from `https://komodoproxy.com` to the one specified in the `outbound_route`.

Lets not over complicate the things and just provide target url separately in the header and just ask to send GET proxy request to simple uri.

In addition I think current implementation in prod cant handle proxy routes properly as in prod code to find outbound by inbound the `uri` is used
https://github.com/KomodoPlatform/komodo-defi-proxy/blob/73738f08bcff7fa3400615c2b9eca9c82f8f0325/src/net/http.rs#L262-L276

However in tests `path` is used
https://github.com/KomodoPlatform/komodo-defi-proxy/blob/73738f08bcff7fa3400615c2b9eca9c82f8f0325/src/net/http.rs#L345-L360

Reply:
> Not the same in our case, even in your example you are adding additional path segment as "/ETH".
Due to inbound/outbound routes GET requests dont work directly. It will work for POST yes.


It should be same in any case. That's the nature of proxy servers. This isn't an API, this is proxy server with some security logic in-between the network operations. You should not change anything on the request, you should not perform anything extra in the data. Just send the request as it is and that's it. All KDP needs to know the signed message in the header and that should be same for every service.

> even in your example you are adding additional path segment as "/ETH".

What do you mean? This is inbound route, we don't change anything in the request. That path supposed to be included in the coin activation.

Reply:
> Lets not over complicate the things and just provide target url separately in the header and just ask to send GET proxy request to simple uri (eg https://komodoproxy.com/nft-proxy).

Current implementation is complicating it, right not proxy server doesn't really act like a proxy server :)

Reply:
> All KDP needs to know the signed message in the header and that should be same for every service.

That what Im going to do.
Right now Im trying to explain the differences between two implementation approaches which manages uris in GET requests differently.

> Current implementation is complicating it, right not proxy server doesn't really act like a proxy server :)

it acts like proxy service, you send request to proxy which targets feature endpoints.

> What do you mean? This is inbound route, we don't change anything in the request.

You have to change url in GET request to proxy layer.

Ive provided example how request will look like and handled in your case suggestion. Do you want to send GET requests to such uri below:

> 1. **KDF Sends Request**:
>    - KDF sends a GET request to `https://komodoproxy.com/nft-proxy/api/  v2.2/0x1f9090aaE28b8a3dCeaDf281B0F12828e676c326/nft/transfers?chain=eth&format=decimal&order=DESC` with the `X-Auth-Payload` header.

Are you ok with extracting first segment path and go trough vector of routes to find `ProxyRoute` on KDP? Then proxy will have to remove this first segment. 
In addition on KDF side:
- if `proxy_auth` is true, then: add `/nft-proxy` segment to path -> then add the rest of the path + query params
- if false add the path and query params as usual.

The above is over complicating thing, while it is possible to simplify process on both sides.

If we provide target url in the X target url header, then we dont need to adding `/nft-proxy`  segment if if `proxy_auth` is true, and removing it on KDP. In both `proxy_auth` true/false cases uri will be constructed in the same way, the only differences is in Headers on KDF side. On proxy side just change base url and remove headers.

Please choose the implementation approach: 
- add (on KDF) and remove (on KDP) inbound segment path in target Url for GET requests
- keep target url in Header

Reply:
I think I repeat myself over and over here.